File: frontend/miniapp/src/api/portal.ts
import { ApiClient } from "./client";

export type LoyaltyDto = {
  balance: number;
  patient_funds: number;
  bonus_funds: number;
  patient_debt: number;
  patient_debt_company: number;
  current_discount?: number | null;
  progress?: number | null;
};

export type ServiceCategoryDto = { id: string; title: string; services_count: number };
export type ServiceDto = { service_id: string; title: string; price?: number | null; category_id?: string | null };
export type AppointmentSlotDto = {
  schedule_id: string;
  doctor_name: string;
  profession?: string | null;
  clinic_name?: string | null;
  room?: string | null;
  date: string;
  time_start: string;
};

export class PortalApi {
  constructor(private readonly client: ApiClient) {}
  loyalty(): Promise<LoyaltyDto> { return this.client.get("/patient/loyalty"); }
  serviceCategories(): Promise<ServiceCategoryDto[]> { return this.client.get("/patient/services/categories"); }
  services(categoryId?: string): Promise<ServiceDto[]> {
    const query = categoryId ? `?category_id=${encodeURIComponent(categoryId)}` : "";
    return this.client.get(`/patient/services${query}`);
  }
  searchServices(q: string, categoryId?: string): Promise<ServiceDto[]> {
    const params = new URLSearchParams({ q });
    if (categoryId) params.set("category_id", categoryId);
    return this.client.get(`/patient/services/search?${params.toString()}`);
  }
  schedule(): Promise<AppointmentSlotDto[]> { return this.client.get("/patient/appointments/schedule"); }
}

File: src/presentation/patient_api/schemas/portal.py
from pydantic import BaseModel


class PatientBalanceResponse(BaseModel):
    balance: float
    patient_funds: float
    bonus_funds: float
    patient_debt: float
    patient_debt_company: float
    current_discount: float | None = None
    progress: float | None = None


class ServiceCategoryResponse(BaseModel):
    id: str
    title: str
    services_count: int
    children: list["ServiceCategoryResponse"] = []


class ServiceResponse(BaseModel):
    service_id: str
    title: str
    code: str | None = None
    price: float | None = None
    category_id: str | None = None
    category_title: str | None = None
    category_path: str | None = None


class AppointmentSlotResponse(BaseModel):
    schedule_id: str
    doctor_id: str | None = None
    doctor_name: str
    profession: str | None = None
    clinic_id: str | None = None
    clinic_name: str | None = None
    room: str | None = None
    date: str
    time_start: str
    time_end: str | None = None
    service_id: str | None = None
